### PR TITLE
Extend and relax TensorList sample APIs

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Add ResizeSample that allows to resize individual samples. Automatically converts the batch into non-contiguous mode, same as SetSample.

SetSample no longer requires the order between the batch and new sample to match - instead it waits for the incoming order in the order of current batch, so the access in the batch order will be correct.

Fix a problem that prohibited copying from empty batch that had no samples nor type set. Now it resets the current batch to similar state.

Those features are useful for implementing the Split and Merge operators.

### Affected modules and functionalities:
TensorList 

### Key points relevant for the review:
If the parameters to order::wait are in correct order.

### Tests:
- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/pipeline/data/tensor_list.cc [hunk 1]
```diff
@@ -573,6 +585,25 @@ void TensorList<Backend>::Resize(const TensorListShape<> &new_shape, DALIDataTyp
...
+  DALI_ENFORCE(IsValidType(type()),
+               "Sample in TensorList cannot be resized with invalid type. Set the type first for "
+               "the whole TensorList using set_type or Resize.");
+  DALI_ENFORCE(sample_dim() == new_shape.sample_dim(),
+               "Sample in TensorList cannot be resized with non-compatible batch dimension. Use "
+               "set_sample_dim or Resize to set correct sample dimension for the whole batch.");
+  // Bounds check
+  assert(sample_idx >= 0 && sample_idx < curr_num_tensors_);
+  // Resizing any individual sample converts the batch to non-contiguous mode
+  MakeNoncontiguous();
```
Comment (mzient) [inline dali/pipeline/data/tensor_list.cc:599]:
```suggestion
  if (volume[new_shape] != volume[shape_.set_tensor_shape(sample_idx)])
    MakeNoncontiguous();
```
?

Comment (klecki) [reply dali/pipeline/data/tensor_list.cc:599]: As discussed on slack, I think if we want to have a variant that doesn't break the contiguity, it should be a separate call like ReshapeSample or ResizeSample with some parameter, that would always enforce that we have a valid volume. That way the bahaviour is consistent and we know the postconditions of the call always.

File: dali/pipeline/data/tensor_list.cc [hunk 2]
```diff
@@ -573,6 +585,25 @@ void TensorList<Backend>::Resize(const TensorListShape<> &new_shape, DALIDataTyp
...
+  if (tensors_[sample_idx].capacity() >= volume(new_shape) * type_.size())
+    return;
+  shape_.set_tensor_shape(sample_idx, new_shape);
```
Comment (stiepan) [inline dali/pipeline/data/tensor_list.cc:602]: Isn't the early exit too early? Shouldn't we remember the new shape even if the actual tensor allocation is not affected?

File: dali/pipeline/data/tensor_list.cc [hunk 3]
```diff
@@ -821,6 +852,13 @@ void TensorList<Backend>::Copy(const TensorList<SrcBackend> &src, AccessOrder or
                                bool use_copy_kernel) {
   auto copy_order = copy_impl::SyncBefore(this->order(), src.order(), order);
 
+  if (!src.has_data() && !IsValidType(src.type())) {
+    Reset();
+    SetLayout(src.GetLayout());
+    // no copying to do
```
Comment (stiepan) [inline dali/pipeline/data/tensor_list.cc:858]: Why not just `if(!src.has_data())`?

Comment (klecki) [reply dali/pipeline/data/tensor_list.cc:858]: We can have no allocation but still have a valid type, in that case reset doesn't work that well.

File: dali/pipeline/data/tensor_list.cc (hunk 2 above)
Comment (klecki) [reply dali/pipeline/data/tensor_list.cc:602]: Yup, thanks for spotting that.

Comment (stiepan) [reply dali/pipeline/data/tensor_list.cc:602]: How about adding a test for such case if that is not much of a hassle?

File: dali/pipeline/data/tensor_list.cc (hunk 3 above)
Comment (klecki) [reply dali/pipeline/data/tensor_list.cc:856]: So if the source has a valid type, I want to copy it, I leave it to the Resize(<empty_shape>, type) to handle.

**Review by stiepan — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Pull request merged**